### DISCUSSION TITLE
Saving the complete Device configuration to a device-specific Config file

### DISCUSSION THREAD

**Discussion by CV-Private** [Bepacom-Raalte/bepacom-HA-Addons#43] [Ideas]
It would be great if I could save the entire configuration of my BACnet device to a file, so that I can restore the settings from a file, or share it with other users who have the same device.

Use-case description: 

I have a Brötje Mono K 11 heatpump that is retrofitted with a GTW-21 BACnet interface module. The BACnet module is universal for many different devices, made by the BDR Thermea group. It has 1279 objects, but only a small subset of the object is used for controlling and monitoring my heatpump.

I have to configure all subscriptions and entities that are being supported by the Mono K11 device, while all unused objects of the GTW-21 can deactivated and do not need to be used in HA.

Configuring the GTW-21 device is a lot of work and I'd like to be able to store the configuration of all entities and subscriptions for the Mono K11 device. This is currently not possible in the Bepacom Add-on. 

The proposal is therefore to add a feature "Save the complete Add-On & Integration config to a file", as well as "Load configuration file".

**Comment by GravySeal** [comment]
I've entertained the idea of creating a config file for a while, but in essence, the configuration of the add-on is already a config file.
It would be a little clumsy to work with multiple locations where the add-on can be configured.

The add-on is made to be sort of "fire and forget", meaning you start it and it just works. For small to medium sites with only a few BACnet devices and points, this is the best. For large sites or BACnet devices with a huge number of objects, some more control is most definitely wanted. For this, I want to propose the following:

- A whitelist for CoV subscriptions, meaning a list has to be made in the add-on configuration of all the objects that are going to be subscribed to. This would prevent the add-on from subscribing to each and every supported object it detects, which is especially useful when only 20 subscriptions can be made.
- A fast and medium poll list, meaning a list of devices with objects that have to be polled at a different rate than the "update interval" configuration. A fast poll can be between 5-60 seconds, medium between 60-600, and update interval will be 600 or greater. Every object will still be read by the update interval, but objects included in the fast or medium poll list will be read as frequently as configured. This gives more control to the user and prevents network congestion. This also addresses [this discussion](https://github.com/Bepacom-Raalte/bepacom-HA-Addons/discussions/41).

Something useful to mention is, the integration and add-on work separately from each other. The add-on arranges all the BACnet stuff while the integration handles the Home Assistant stuff by gathering the data from the add-on. Disabling entities by default is already an option when setting up the integration for the first time, it just won't change anything anymore when trying to reconfigure the integration.

**Comment by CV-Private** [comment]
Would it be possible and in accordance with the BACnet standard to use the EDE file for this purpose? For example, I could create an EDE file specifically for my heatpump model, which would only contain the objects that are needed for monitoring and controlling that device.
The EDE file could be loaded to the Add-On, which would then only enumerate the objects that are contained in the EDE. All other objects would be ignored.
The integration would then only create the entities from the enumerated objects.

Maybe this is already "somehow" foreseen in the Add-On, but I have currently no clue how those menu items are supposed to work and what is meant with "Select State Texts". If that is documented anywhere, could you please point me to a description?

![image](https://github.com/Bepacom-Raalte/bepacom-HA-Addons/assets/108861703/a74de2ac-73bf-44cd-ac7d-d613cc6505c7)

**Comment by GravySeal** [comment]
If you download v1.4.1b1, you'll see there's some new configuration options:

![image](https://github.com/Bepacom-Raalte/bepacom-HA-Addons/assets/93523925/843ac91d-7331-453c-ba89-36c2b4ffd049)

CoV whitelist isn't implemented yet.

I think modifying the EDE file and then using it for entities is a little odd. It's meant as a placeholder until the real devices joins the network.
I think a better solution would be to make an option in the add-on to prevent scanning any device in the network, and instead configure it all in the configuration. This way, no entities would be created for objects you don't want to use. 
The current way it works is everything the add-on saved will be sent to Home Assistant to use as data. Can't really make exceptions in that without basically rebuilding everything.

**Reply by CV-Private** [reply]
Got it, thanks! I just installed v1.4.1b1 and will play with it.
May I suggest that you add units to the field descriptions, for example "Fast poll rate **[s]**"? That would save the user from guessing...

Happy Easter!

**Reply by CV-Private** [reply]
Unfortunately, I can't get the new features working yet.

The subscriptions to the objects were not automatically made. The subscription list in the Add-On and in Yabe was empty and I had to manually subscribe to the objects that I need.

The polling is also not working for me. The parameters are not getting updated, neither at 5s nor 300s rate.

I noticed an error at the beginning of the Add-On logfile: [bacnet_addon_logs.zip](https://github.com/Bepacom-Raalte/bepacom-HA-Addons/files/14813283/bacnet_addon_logs.zip)

> [16:38:29]|DEBUG    |main.py           ->main                                : [{'objects': ['analogValue:258', 'analogValue:259', 'analogValue:1673']}]
> [16:38:29]|ERROR    |main.py           ->exception_handler                   : **An uncaught error occurred:** 
> NoneType: None

My configuration is as follows. I also tried _- devicename: device:783351856_, but with no difference is the result:

![image](https://github.com/Bepacom-Raalte/bepacom-HA-Addons/assets/108861703/5e5ff88c-e7da-4efe-af8d-b8303df2f80f)

Am I making something wrong?

**Reply by GravySeal** [reply]
Try deviceName instead of devicename. Also CoV whitelist isn't implemented yet so that option won't do much. 
I am planning on making a larger structure for all these options in the next release.

**Comment by CV-Private** [comment]
I restarted the Add-On with _deviceName_ in the config. Apparently it is polling now.

However, unlike CoV subscriptions, I'm seeing the Entities flashing as "not available" during the update of the presentValues. 

See how it looks every 5s, which is my FastPoll interval: 

https://github.com/Bepacom-Raalte/bepacom-HA-Addons/assets/108861703/fa355cca-b0be-4111-931b-bfdc1c5722eb


It looks even worse when I use graphical cards like guages:

https://github.com/Bepacom-Raalte/bepacom-HA-Addons/assets/108861703/413e716f-8784-4b03-aae8-37e9371d7ab7

**Reply by GravySeal** [reply]
The flashing might be because the integration is reloading as the data received is increasing in size with new objects. 
I accidentally left the timer for that at 3 seconds instead of 30. Does it still flash now or not?
If so, please share some logs.

**Comment by CV-Private** [comment]
Unfortunately, the poll task stopped and the log is full of errors. The last update happened 2h ago. :-(

![image](https://github.com/Bepacom-Raalte/bepacom-HA-Addons/assets/108861703/8fa144d0-e198-4a1c-9443-8015faa113ba)

Here are the Add-On and Integration logs:

[bacnet_addon_logs_2024-03-31.zip](https://github.com/Bepacom-Raalte/bepacom-HA-Addons/files/14817451/bacnet_addon_logs_2024-03-31.zip)
[home-assistant_bacnet_interface_2024-03-31T19-37-52.745Z.zip](https://github.com/Bepacom-Raalte/bepacom-HA-Addons/files/14817452/home-assistant_bacnet_interface_2024-03-31T19-37-52.745Z.zip)

**Reply by GravySeal** [reply]
Actually it doesn't seem that crazy. The fast polling kept polling, only the slow poll does nothing.
CoV doesn't work yet, so the config for that doesn't do anything. If you look at analogValue:1673, did that update in the meantime?

**Comment by CV-Private** [comment]
The AV1673 is a mode switch that won't change unless I change it manually. I therefore included other objects to the FastPoll list that do change dynamically. Unfortunately, I don't see the values getting updated in the Entities. The last update of all those values was 10h ago, which is when I restarted HA:

![image](https://github.com/Bepacom-Raalte/bepacom-HA-Addons/assets/108861703/98b0b330-dfc9-4f7a-a16f-e5022964ad89)

![image](https://github.com/Bepacom-Raalte/bepacom-HA-Addons/assets/108861703/1f3467be-97c4-4b5f-aa6e-bdac723807be)

The log file only tells me that "some" value was updated, but not what the object and latest presentValue was. It would perhaps be useful to see that in the Debug log. Currently, it says only:

> 2024-03-31 23:47:19.878 DEBUG (MainThread) [custom_components.bacnet_interface] Manually updated bacnet_interface data

Here are two logs that I captured last night:

[home-assistant_bacnet_interface_2024-03-31T21-47-22.800Z.zip](https://github.com/Bepacom-Raalte/bepacom-HA-Addons/files/14821775/home-assistant_bacnet_interface_2024-03-31T21-47-22.800Z.zip)

[bacnet_addon_logs_2024-03-31b.zip](https://github.com/Bepacom-Raalte/bepacom-HA-Addons/files/14821776/bacnet_addon_logs_2024-03-31b.zip)

**Reply by GravySeal** [reply]
Could you try dev version v1.4.1b2? I just updated with some new config options. Check the documentation to see how it's set up.

**Comment by CV-Private** [comment]
The polling doesn't work yet for me. This is my configuration:

![image](https://github.com/Bepacom-Raalte/bepacom-HA-Addons/assets/108861703/0e230c3a-2d6c-4511-980e-49045b171d33)

The presentValues were not getting updated by FastPoll for the last 2h, since I installed v1.4.1b2.

![image](https://github.com/Bepacom-Raalte/bepacom-HA-Addons/assets/108861703/7821ab55-346d-4938-96bb-4be5d810a284)

Here are the log files:

[bacnet_addon_logs_2024-04-01.zip](https://github.com/Bepacom-Raalte/bepacom-HA-Addons/files/14828187/bacnet_addon_logs_2024-04-01.zip)

[home-assistant_bacnet_interface_2024-04-01T20-09-21.668Z.zip](https://github.com/Bepacom-Raalte/bepacom-HA-Addons/files/14828188/home-assistant_bacnet_interface_2024-04-01T20-09-21.668Z.zip)


I'm going to test it again with SlowPoll removed in the config. I'm suspecting that 600s are not enough to poll "all" values, so that the poll task might get screwed up. I'm seeing many errors related to the Poll task in the logs...

**Comment by GravySeal** [comment]
Found the issue with CoV, that should work now. For the slow poll, try it with an empty slow_poll_list. See if the quick poll and CoV update at least. Your device really does take a long time to read everything.
Dev version v1.4.1b3 should work.
Almost all errors in the logs are because none of the objects support the reliability property, so that's not too important.

**Comment by CV-Private** [comment]
I'll install v1.4.1b3 now. The empty slow_poll_list with v1.4.1b2 didn't help.

**Comment by CV-Private** [comment]
Still no success, even with v1.4.1b3, unfortunately.

The CoV list in the Add-On remains empty and the Entity values are not getting updated, neither by CoV nor QuickPoll, after the Add-On has finished reading all objects.

This is my current config:

![image](https://github.com/Bepacom-Raalte/bepacom-HA-Addons/assets/108861703/ddea8d84-dd31-46ac-8a2c-81c44f13752d)

I noticed from an error in the log that I made a mistake. It actually needs to be - analog**Value**:459. I corrected it and restarted the Add-On.

Otherwise, the Poll task attempts to read the QuickPoll objects every 5s, but the Entities are not getting updated.

While the Object list is being read at startup, I'm still seeing the "not available" flashing each ~3s. Shouldn't the timer already be at 30s in v1.4.1b3?

Here are the logs (still with the error about non-existing object - analogInput:459):

[bacnet_addon_logs_2024-04-01b.zip](https://github.com/Bepacom-Raalte/bepacom-HA-Addons/files/14829064/bacnet_addon_logs_2024-04-01b.zip)
[home-assistant_bacnet_interface_2024-04-01T22-17-06.320Z.zip](https://github.com/Bepacom-Raalte/bepacom-HA-Addons/files/14829065/home-assistant_bacnet_interface_2024-04-01T22-17-06.320Z.zip)

**Comment by Bepacom-Raalte** [comment]
Can you try to include slow_poll_list with slow_poll_rate in your configuration? Just make it empty like so: 
`slow_poll_list: []`
`slow_poll_rate: 600`

> Shouldn't the timer already be at 30s in v1.4.1b3?

This is handled by the integration. You should see the add-on as a separate piece of software running beside Home Assistant, while the integration is part of Home Assistant.
v1.4.1b4 should see your entities being updated though, either by CoV or polling.

**Comment by CV-Private** [comment]
Good news! With v1.4.1b4 and the SlowPoll in the config, the entities are getting updated and the CoV subscriptions have been made:

![image](https://github.com/Bepacom-Raalte/bepacom-HA-Addons/assets/108861703/4bb7d118-cccf-4566-888f-880d6f65c10b)

![image](https://github.com/Bepacom-Raalte/bepacom-HA-Addons/assets/108861703/ae5c35fd-54a3-4e5b-8854-1222a6f4476b)

Here's the Add-On log. (Debug logging was off in my latest test):

[bacnet_addon_logs_2024-04.01c.zip](https://github.com/Bepacom-Raalte/bepacom-HA-Addons/files/14833157/bacnet_addon_logs_2024-04.01c.zip)


All Entities are getting updated after some seconds:

![image](https://github.com/Bepacom-Raalte/bepacom-HA-Addons/assets/108861703/bdef9809-be1e-424d-97d5-5c85d64d087a)

I noticed that sometimes the values are being refreshed every few seconds, but sometimes it can can quite a bit longer than the configured 5s update interval. 
May I assume that the Entity is counting the time until the presentValue actually **changes**? For example, if the pump speed would be constant for 1minute, the Entity would count the time up to a minute, even though the property is getting polled every 5s?
It would make sense to me - just trying to understand if the behavior meets the expectation.

**Reply by Bepacom-Raalte** [reply]
Great! Glad it seems to work now.

> For example, if the pump speed would be constant for 1minute, the Entity would count the time up to a minute, even though the property is getting polled every 5s?

This is correct, if the value remains the same, it won't show as updated. 
The add-on sends data to the integration each update, so even though it doesn't show it in Home Assistant, the values should always be the latest data from the add-on.
